### ISSUE TITLE
"Specified cast is not valid"

### ISSUE THREAD

**Issue by MarcosCobena** [JoeMayo/LinqToTwitter#69]
Hi Joe,

I've been using your client for a personal project which's currently published in Android and everything's perfect. However, I'm facing a few issues with iOS version.

I had initially an issue related to a malformed IL coming from the PCL version of the library. Updating today such to 4.1.0 I'm getting this other exception: "Specified cast is not valid."

  at Microsoft.Scripting.Interpreter.EqualInstruction+EqualInt64.DoCalculate (System.Object l, System.Object r) [0x00000] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/EqualInstruction.cs:84 
  at Microsoft.Scripting.Interpreter.ComparisonInstruction.Calculate (System.Object l, System.Object r) [0x00015] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/ComparisonInstruction.cs:44 
  at Microsoft.Scripting.Interpreter.ArithmeticInstruction.Run (Microsoft.Scripting.Interpreter.InterpretedFrame frame) [0x00020] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/ArithmeticInstruction.cs:45 
  at Microsoft.Scripting.Interpreter.Interpreter.Run (Microsoft.Scripting.Interpreter.InterpretedFrame frame) [0x0001b] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Interpreter.cs:126 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00047] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:357 
  at BuenosDias.iOS.BuenosDiasPicturesProvider+<RetrieveTweetsAsync>c__async4.MoveNext () [0x00353] in /Users/mcobena/Repositorios/BuenosDias/BuenosDias.iOS/ViewControllers/PicturesGalleryViewController.cs:182 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00047] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:357 
  at BuenosDias.iOS.BuenosDiasPicturesProvider+<RetrievePicturesAsync>c__async0.MoveNext () [0x00067] in /Users/mcobena/Repositorios/BuenosDias/BuenosDias.iOS/ViewControllers/PicturesGalleryViewController.cs:80 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00047] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter`1[TResult].GetResult () [0x00000] in /Users/builder/data/lanes/3412/3cf8aaed/source/maccore/_build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/external/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:357 
  at BuenosDias.iOS.PicturesGalleryViewController+<ViewDidLoad>c__async0.MoveNext () [0x00086] in /Users/mcobena/Repositorios/BuenosDias/BuenosDias.iOS/ViewControllers/PicturesGalleryViewController.cs:26 

That last line is in the way of:

```
                var tweetsQuery = from tweet in twitterContext.Status
                                  where tweet.Type == StatusType.User &&
                                  tweet.ScreenName == "reinoso_alvaro" &&
                                  tweet.Text.Contains("#buenosdias") &&
                                  tweet.IncludeRetweets == false &&
                                  tweet.TrimUser == true &&
                                  tweet.ExcludeReplies == true &&
                                  tweet.SinceID == InitialTweetId &&
                                  tweet.Count == MaximumTweetsToRetrieve
                                  select tweet;
                var first = tweetsQuery.FirstAsync();
```

where trying to materialize the query returns that exception.

I can provide any other information you might need if you need it.

**Thanks** in advance for your work with this library, I really like it and find it helpful,

Marcos

**Comment by JoeMayo** [comment]
Hi Marcos,

Does it help if you await the materialization, like this:

    var first = await tweetsQuery.FirstAsync();

LINQ to Twitter is async, so you'll need your call chain to be async too.

**Comment by MarcosCobena** [comment]
Thanks Joe,

Sorry it doesn't help. My fault I pasted here the version without async, but initially I've been doing a ToListAsync() to the query, awaiting it correctly, and that's the version working currently in Android with 3.1.2.

**Comment by MarcosCobena** [comment]
Hi again,

After referencing directly the source code and having the same behavior, I've been playing around with commenting lines from the query and the issue doesn't exist anymore if I comment these lines:

```
tweet.SinceID == InitialTweetId &&
tweet.Count == MaximumTweetsToRetrieve
```

I'll keep researching to see what happens. Thanks!

**Comment by JoeMayo** [comment]
What are the values that you have for InitialTweetId and Count. The max for Count is 200. Also, setting SinceID to 0 will cause an exception that looks like the one you have above. Setting SinceID to 1 should work.

**Comment by MarcosCobena** [comment]
The issue is definitely related to this line:

`tweet.SinceID == InitialTweetId`

Since InitialTweetId is declared as ulong with value 1, why then the exception mentions the comparison between Int64? The exact line where the error happens is

[https://github.com/mono/mono/blob/master/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/EqualInstruction.cs#L84](https://github.com/mono/mono/blob/master/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/EqualInstruction.cs#L84)

```
        internal sealed class EqualInt64 : EqualInstruction {
            protected override object DoCalculate (object l, object r)
            {
                return (Int64)l == (Int64)r; // Exception raises here, where l and r should be ulong initially
            }
        }
```

Shoudln't it be using the [EqualUInt64](https://github.com/mono/mono/blob/master/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/EqualInstruction.cs#L109) equivalent?

**Comment by JoeMayo** [comment]
SinceID is ulong. What type is InitialTweetID?

**Comment by MarcosCobena** [comment]
ulong as well

**Comment by jderrough** [comment]
Hi @MarcosCobena,

I'm having the exact same problem, did you manage to make it work?

**Comment by jderrough** [comment]
I did a bit of digging and it seems like the comparers for unsigned integers in mono up to version 4.8 are bound to their signed counterpart (https://github.com/mono/mono/blob/mono-4.8.0-branch/mcs/class/dlr/Runtime/Microsoft.Dynamic/Interpreter/Instructions/EqualInstruction.cs):

```
case TypeCode.UInt16: return _UInt16Lifted ?? (_UInt16Lifted = new EqualInt16() { LiftedToNull = true });
case TypeCode.UInt32: return _UInt32Lifted ?? (_UInt32Lifted = new EqualInt32() { LiftedToNull = true });
case TypeCode.UInt64: return _UInt64Lifted ?? (_UInt64Lifted = new EqualInt64() { LiftedToNull = true });
```

I thought this might be the ultimate cause of the exception, but the following test runs fine:

```
var ulongList = new List<ulong> { 1ul, 2ul, 3ul };
var ones = (from u in ulongList where u == 1ul select u).ToList();
```

My understanding of the nooks and crannies of LinqToTwitter and the Mono APIs involved is very limited so any help would be definitely appreciated.

**Comment by jderrough** [comment]
Well, casting the Status.SinceID property did the trick:

```
[...]
&& (long)tweet.SinceID == sinceStatusId // sinceStatusId is a long
[...]
````

**Comment by jsauve** [comment]
Was suffering from the same problem, but with the `MaxID` property, which is also a `ulong`.

I wound up just casting the `ulong` values in my query to `long`, and everything works great.

```
var subsequentResults = 
    (await
        (from search in twitterContext.Search
	where
	search.Type == SearchType.Search &&
	search.Count == count &&
	search.ResultType == ResultType.Mixed &&
	search.IncludeEntities == true &&
	search.Query == query &&
	(long)search.MaxID == (long)lowestId.Value - 1  // cast ulong to long to workaround InvalidCastExeption originating from Mono equality comparer for Int64
	select search) .SingleOrDefaultAsync())?.Statuses;
```

**Comment by JoeMayo** [comment]
I believe this is working now. I pushed a beta to NuGet: 

https://www.nuget.org/packages/linqtotwitter/4.2.0-beta3

Let me know if you see any more problems. I'm going through the issues list to see if there are any other unresolved items and will probably push a full version soon.

**Comment by JoeMayo** [comment]
This was fixed in [LINQ to Twitter v4.2.0](https://github.com/JoeMayo/LinqToTwitter/releases/tag/v4.2.0).

[@JoeMayo](https://twitter.com/JoeMayo)
